### DISCUSSION TITLE
자료구조 - 그래프 - 서희 & 재찬

### DISCUSSION THREAD

**Discussion by LireEruel** [Shortudy-10th/tech-for-developer#68] [질의 응답]
**❓️ 
아래 2가지 질문에 답해주세요.
그래프 IS A 트리. 참 거짓여부와 그 이유
트리 IS A 그래프. 참 거짓여부와 그 이유**

**💬 답변**

**❓️프로그램상에서 그래프를 구현한다면, 어떤 방법으로 구현할 수 있을까요?**

**💬 답변**

**❓️대표적인 그래프 탐색법 두 개에 대해 간략하게 말해주세요.**

**💬 답변**

**❓️완전 그래프와 연결 그래프의 차이점은 무엇일까요?**

**💬 답변**

**❓️최소신장트리에 대해 설명해주세요.**

**💬 답변**

**❓️다익스트라 알고리즘과 벨만포드 알고리즘에 대해 설명하고 두 알고리즘의 차이점을 설명해주세요**

**💬 답변**

**❓️인접 행렬과 인접 리스트의 장단점은 무엇인가요?**

**💬 답변**

**Comment by LireEruel** [comment]
**❓️ 
아래 2가지 질문에 답해주세요.
그래프 IS A 트리. 참 거짓여부와 그 이유
트리 IS A 그래프. 참 거짓여부와 그 이유**

**Reply by kjc6735** [reply]
그래프란 노드와 노드 간을 연결하는 간선으로 구성된 자료 구조이다.
이를 통해 연결된 노드 간의 관계를 표현할 수 있는 자료 구조이다. 
특징으로는 순환, 비순환 구조를 이룬다. 방향성이 있는 그래프와 방향성이 없는 그래프가 있다.

트리는 노드와 노드 간을 연결하는 간선으로 구성된 자료구조이다.
트리는 두 개의 노드 사이에 반드시 1개의 경로만 가지기 떄문에 싸이클이 존재하지 않는다.

그래프 IS A 트리 =>  거짓
그래프는 트리를 포함한 더 큰 개념이며, 그래프의 특수한 형태 중 하나이다.

트리 IS A 그래프 => 참
트리는 그래프의 종류 중 하나이므로 (포함되므로) 트리는 그래프가 맞다.

**Comment by LireEruel** [comment]
**❓️프로그램상에서 그래프를 구현한다면, 어떤 방법으로 구현할 수 있을까요?**

**💬 답변**

**Reply by kjc6735** [reply]
2차원 배열
2차원 배열에 모든 정점의 연결 여부를 표현한다. 이 방법은 간선의 존재 여부를 쉽게 알 수 있지만, 연결되지 않은 것들도 포함하여 나타내기 때문에 메모리 낭비가 심하다. 

인접 리스트
노드에 연결된 정점들을 리스트에 저장하는 방법이다. 간선의 추가 및 제거가 빠르고 메모리 사용량이 적다.

간선 리스트
그래프의 간선을 리스트로 표현하는 방법이다. 각 간선을 구조체나 객체로 나타내고, 리스트에 저장한다.

**Comment by LireEruel** [comment]
**❓️대표적인 그래프 탐색법 두 개에 대해 간략하게 말해주세요.**

**💬 답변**

**Reply by kjc6735** [reply]
깊이 우선 탐색
 한 방향으로 갈 수 있을 때 까지 가다가 더 이상 갈 수 없게 되면 가장 가까운 갈림길로 돌아와서 다시 다른 방향으로 다시 탐색

넓이 우선 탐색
 가장 가까운 노드들 부터 탐색하는 방법.

**Comment by LireEruel** [comment]
**❓️완전 그래프와 연결 그래프의 차이점은 무엇일까요?**

**💬 답변**

**Reply by kjc6735** [reply]
완전 그래프는 각 정점에서 다른 모든 정점을 연결하여 가능한 최대의 간선 수를 가진 그래프이다. 
연결 그래프는 어떤 두 정점 사이에 경로가 존재하는 것이다. 

완전 그래프는 노드가 다른 노드로 갈 때 직접 연결되어 있어 바로 갈 수 있다.
하지만 연결 그래프는 두 노드가 직접 연결되어 있지 않을 수 있고, 다른 노드를 통해 갈 수 있다.

**Comment by LireEruel** [comment]
**❓️최소신장트리에 대해 설명해주세요.**

**💬 답변**
## 최소 신장 트리(Minimum Spanning Tree, MST)

- 그래프 내의 모든 정점을 연결하면서 간선의 가중치 합이 최소가 되는 트리
- 그래프의 모든 정점을 포함하고 싸이클이 없는 연결 선분의 집합

**Reply by LireEruel** [reply]
**❓️최소 신장 트리를 찾는 알고리즘을 설명해보세요.**

**💬 답변**
### 1. Kruskal 알고리즘
 
- 간선을 가중치의 오름차순으로 정렬합니다.
- 가장 작은 가중치를 가진 간선부터 선택하면서 트리를 확장합니다.
- 싸이클이 형성되는 간선은 제외합니다. (Union-Find 알고리즘 사용)
- 모든 정점이 연결될 때까지 반복합니다.

### 2. Prim 알고리즘

- 시작 정점을 선택합니다.
- 선택한 정점과 인접한 간선 중 최소 가중치를 가진 간선을 선택합니다.
- 해당 간선에 연결된 정점을 트리에 추가합니다.
- 모든 정점이 트리에 포함될 때까지 반복합니다.

**Comment by LireEruel** [comment]
**❓️다익스트라 알고리즘과 벨만포드 알고리즘에 대해 설명하고 두 알고리즘의 차이점을 설명해주세요**

**💬 답변**

## 다익스트라 알고리즘 (Dijkstra's Algorithm)
다익스트라 알고리즘은 시작 정점에서 다른 모든 정점까지의 최단 경로를 구하는 알고리즘입니다. 음의 가중치를 가진 간선이 없는 경우에 사용할 수 있습니다.

1. 시작 정점을 선택하고, 해당 정점과의 거리를 0으로 초기화하고, 다른 모든 정점과의 거리를 무한대로 초기화합니다.
2. 현재 정점에서 연결된 인접 정점들 중 아직 방문하지 않은 정점의 거리를 업데이트합니다.
3. 아직 방문하지 않은 정점 중 최단 거리를 가진 정점을 선택하고, 그 정점을 현재 정점으로 설정한 후 2번 과정을 반복합니다.
4. 모든 정점을 방문할 때까지 2-3번 과정을 반복합니다.

## 벨만-포드 알고리즘 (Bellman-Ford Algorithm)
벨만-포드 알고리즘은 시작 정점에서 다른 모든 정점까지의 최단 경로를 구하는 알고리즘으로, 음의 가중치를 가진 간선이 있어도 사용할 수 있습니다.

1. 시작 정점을 선택하고 해당 정점과의 거리를 0으로 초기화하고, 다른 모든 정점과의 거리를 무한대로 초기화합니다.
2. 전체 간선을 순회하면서, 간선의 시작 정점까지의 거리와 해당 간선의 가중치의 합이 간선의 끝 정점까지의 현재 거리보다 작으면 해당 거리를 업데이트합니다.
3. 2번 과정을 정점의 수 - 1번 반복합니다.
4. 음의 사이클 존재 여부를 확인하기 위해 한 번 더 모든 간선을 순회하며 거리를 업데이트합니다. 이때 거리가 업데이트되면 음의 사이클이 존재한다고 판단합니다.

## 두 알고리즘의 차이점

- 가중치의 부호: 다익스트라는 음의 가중치가 없는 경우에만 사용 가능하며, 벨만-포드는 음의 가중치가 있어도 사용 가능합니다.

- 시간 복잡도: 다익스트라 알고리즘이 일반적으로 벨만-포드 알고리즘보다 빠릅니다. 다익스트라는 O((V+E)logV), 벨만-포드는 O(VE)의 시간 복잡도를 가집니다.

- 알고리즘 구조: 다익스트라는 우선순위 큐를 사용하여 인접한 정점을 갱신하는 반면, 벨만-포드는 모든 간선을 순회하면서 거리를 갱신합니다.

- 음의 사이클 탐지: 벨만-포드는 음의 사이클의 존재 여부를 판별할 수 있으나, 다익스트라는 그렇지 않습니다.

**Comment by LireEruel** [comment]
**❓️인접 행렬과 인접 리스트의 장단점은 무엇인가요?**

**💬 답변**

## 인접 행렬 (Adjacency Matrix)

### 장점
- 직관적인 표현: 그래프의 연결 상태를 2차원 행렬로 쉽게 표현할 수 있습니다.
- 빠른 연결 확인: 두 노드 간의 연결 유무를 O(1)의 시간 복잡도로 확인할 수 있습니다.

### 단점
- 메모리 낭비: 모든 노드 간의 연결 정보를 저장해야 하므로, 연결이 적은 그래프의 경우 메모리 낭비가 발생할 수 있습니다.
- 접근 시간: 특정 노드와 연결된 모든 노드를 방문하려면 O(V)의 시간이 걸리므로, 노드 수에 비해 간선 수가 적은 경우 비효율적일 수 있습니다.

## 인접 리스트 (Adjacency List)

### 장점
- 메모리 효율: 실제 연결된 노드의 정보만 저장하므로, 메모리를 효율적으로 사용할 수 있습니다.
- 빠른 엣지 순회: 특정 노드와 연결된 모든 노드를 방문하는 데 필요한 시간 복잡도는 O(E)로, 연결된 간선의 수에 비례합니다.

### 단점
- 연결 확인 느림: 두 노드 간의 연결 유무를 확인하려면 해당 노드의 리스트를 전부 확인해야 하므로 O(V)의 시간이 걸릴 수 있습니다.
- 구현 복잡성: 인접 리스트의 구현은 인접 행렬에 비해 약간 더 복잡할 수 있습니다.
